Add TransactionEvent for successful transactions

Chew opened the pull request:
This adds an event for plugins to hook into when a transaction occurs, allowing them to do whatever they want.

Closes #3635

Was also considering making it cancellable, may do that if suggested.

![bruh](https://cdn.discordapp.com/attachments/134445052805120001/751669803621548143/unknown.png)

Chew committed:
Add TransactionEvent for successful transactions
This adds an event for plugins to hook into when a transaction occurs, allowing them to do whatever they want.

Implements #3635

mdcfe commented on Essentials/src/net/ess3/api/events/TransactionEvent.java:
```diff
@@ -0,0 +1,53 @@
+    private final BigDecimal amount;
+
+    public TransactionEvent(CommandSource requester, IUser target, BigDecimal amount) {
+        super();
```
Should probably check the current thread here, in case transactions occur async.

Chew replied:
Not sure what you mean, doesn't the event get raised after the transaction is over?

Chew committed:
Check if thread is primary thread

pop4959 commented on Essentials/src/com/earth2me/essentials/User.java:
```diff
@@ -11,10 +11,7 @@
-import net.ess3.api.events.JailStatusChangeEvent;
-import net.ess3.api.events.MuteStatusChangeEvent;
-import net.ess3.api.events.UserBalanceUpdateEvent;
+import net.ess3.api.events.*;
```
Don't use * import please

pop4959 replied:
```suggestion
import net.ess3.api.events.AfkStatusChangeEvent;
import net.ess3.api.events.JailStatusChangeEvent;	
import net.ess3.api.events.MuteStatusChangeEvent;	
import net.ess3.api.events.UserBalanceUpdateEvent;
```

pop4959 commented on Essentials/src/com/earth2me/essentials/User.java:
```diff
@@ -11,10 +11,7 @@
-import net.ess3.api.events.JailStatusChangeEvent;
-import net.ess3.api.events.MuteStatusChangeEvent;
-import net.ess3.api.events.UserBalanceUpdateEvent;
+import net.ess3.api.events.*;
```
Don't use * imports

Chew replied:
That's just what IntelliJ auto optimize did ;-;

pop4959 replied:
you can turn it off

Octol1ttle commented on Essentials/src/com/earth2me/essentials/User.java:
```diff
@@ -199,6 +196,8 @@ public void payUser(final User reciever, final BigDecimal value, UserBalanceUpda
             reciever.setMoney(reciever.getMoney().add(value), cause);
             sendMessage(tl("moneySentTo", NumberUtil.displayCurrency(value, ess), reciever.getDisplayName()));
             reciever.sendMessage(tl("moneyRecievedFrom", NumberUtil.displayCurrency(value, ess), getDisplayName()));
+            TransactionEvent transactionEvent = new TransactionEvent(this.getSource(), reciever, value);
+            ess.getServer().getPluginManager().callEvent(transactionEvent);
```
It's "receiver"

Chew replied:
No it's not

Chew replied:
To clarify: it's spelt that way in the code, not necessarily the proper spelling. I'm not here to fix typos.

Octol1ttle replied:
My inner perfectionist couldn't resist.

pop4959 replied:
@l1ttleO Feel free to make a separate PR to fix the typos if you feel the need, but it's not in scope for this one.

Chew committed:
Remove wildcard import
Being harassed by wildcard imports? Just say no! IntellJ legally cannot use wildcard imports without your consent!

Chew commented:
Oops

Chew committed:
Merge remote-tracking branch 'original/2.x' into feature/transactionevent

Chew committed:
Move TransactionEvent.java to proper folder

Chew committed:
Fix some CheckStyle errors

Chew committed:
Move TransactionEvent.java to proper folder... again

Chew committed:
Merge remote-tracking branch 'original/2.x' into feature/transactionevent

Chew committed:
Fix TransactionEvent import in User.java
